### ISSUE TITLE
During wp media regenerate, make media deletion optional

### ISSUE THREAD

**Issue by jaredatch** [wp-cli/wp-cli#1712]
When `wp media regenerate --yes` is executed, what happens is a little deceiving to me.

Part of the work flow is running `$this->remove_old_images( $image->ID );` which removes all the old images. This is fine, however IMO this isn't clearly communicated or conveyed in the documentation. At the very least it should be documented and specifically stated that old images are removed. If uses are coming from plugin solutions (such as the popular [Regenerate Thumbnails](https://wordpress.org/plugins/regenerate-thumbnails/)) this might be an un expected behavior.

Ideally a flag would be added where the media can be regenerated, but leaving all exiting media files and sizes untouched, which I feel is more appropriate for most scenarios.

Heres a example I ran into this week that caused problems:

Client has a popular blog from 2007 with over 10,000 images in the media library and thousands of posts. We recently launch a new version of the site and in this new version the archives thumbnails are a custom size along with a few other images uses through out the new design. With out really giving it much thought I run `wp media regenerate --yes`. Three days later I find out the majority of "Large" and "Medium" images inserted into posts are now broken.

What seems to have happened, is upon regeneration, the image algorithm used to resize the images decided to round up instead of down. No idea if this is due to an GD/ImageMagic at some point or what. But what happened is every image size that does not have hard cropping turned on (on this site it is only Large and Medium) is 1 pixel taller that the previous version of the images.

So a post will reference `image-600x253.jpg` when in the uploads directory that file is `image-600x254.jpg`. Most the URLs in the posts for these images are now incorrect and broken.

Had I had an option to preserve the original images and, _in addition to that_, create the new image sizes I would simply have two versions of the images - `image-600x253.jpg` **and** `image-600x254.jpg` - which I would completely be OK with. The alternative I'm facing is much worse.

So TL;DR - Make it more obvious in the [docs](http://wp-cli.org/commands/media/regenerate/) that regeneration deletes images and ideally make a toggle where this can be disabled.

**Comment by danielbachhuber** [comment]
I understand the pain, but I'm not convinced it makes sense to change the default behavior. We could add a `--preserve-images` flag or similar, though.

**Comment by jaredatch** [comment]
I think that would be plenty adequate, for sure. 

I wasn't concerned about the default setting so much as 1) be more clear that images are removed in documentation 2) provide flag so the behavior can be opt out.

I know on most large existing sites I work on, when I come in I have no idea where odd image sizes might be used in posts or linked to in various ways, so I would typically opt to leave older "legacy" images alone.

**Comment by michalkleiner** [comment]
I'd be interested in this flag to preserve old images as well. I might give it a crack and submit a pull request, the change is not that big as I already briefly browsed the code.

**Comment by gerardjp** [comment]
Sorry to butt in, but I was triggered by this charming option, another possibility might be creating a zip containing all that was removed and placing it in wp-contents/upload.

**Comment by jaredatch** [comment]
:+1: thanks Daniel!

**Comment by nickcernis** [comment]
I know this is ancient, but just wanted to mention the reason for the `image-600x253.jpg` file becoming `image-600x254.jpg` after being regenerated – it's likely due to this bug in WP core: https://core.trac.wordpress.org/ticket/18532. (WP used to round image sizes down instead of up.)

That was fixed in WP 4.1, so an image resized in WP 4.0 might be 253px high. If you then update WP to 4.1 or higher and regenerate thumbnails, the same image will then 254px high. I can reproduce this locally:

<img width="884" alt="screen shot 2016-04-13 at 10 49 21" src="https://cloud.githubusercontent.com/assets/647669/14490406/9267729a-016a-11e6-9116-ffdd40e1e63e.png">

It's not a bug in WP CLI, of course, but just wanted to document this in case others encounter it and wonder why.

**Comment by jaredatch** [comment]
That definitely makes sense. Thanks for tracking that down 😀

**Comment by mstoldt** [comment]
Back from the future: Thanks for making this issue clear @nickcernis as I'm about to regenerate 20k images for a high-traffic blog and was concerned about possible errors!
